File: 1286-IteratorforCombination/1286-IteratorforCombination.java
/*

1286. Iterator for Combination
Medium

Design an Iterator class, which has:

    A constructor that takes a string characters of sorted distinct lowercase English letters and a number combinationLength as arguments.
    A function next() that returns the next combination of length combinationLength in lexicographical order.
    A function hasNext() that returns True if and only if there exists a next combination.

Example:

CombinationIterator iterator = new CombinationIterator("abc", 2); // creates the iterator.

iterator.next(); // returns "ab"
iterator.hasNext(); // returns true
iterator.next(); // returns "ac"
iterator.hasNext(); // returns true
iterator.next(); // returns "bc"
iterator.hasNext(); // returns false
 

Constraints:

    1 <= combinationLength <= characters.length <= 15
    There will be at most 10^4 function calls per test.
    It's guaranteed that all calls of the function next are valid.

*/

class CombinationIterator {

    List<String> list = new ArrayList<>();
    int listPos = 0;
    public CombinationIterator(String characters, int combinationLength) {
        dfs(characters, combinationLength, "", 0);
    }
    private void dfs(String s, int len, String curr, int idx){
        if(curr.length() == len){
            list.add(curr);
            return;
        }
        for(int i = idx; i < s.length(); i++){
            dfs(s,len, curr + s.charAt(i), i + 1);
        }
    }
    
    public String next() {
        return list.get(listPos++);
    }
    
    public boolean hasNext() {
        if(listPos == list.size())
            return false;
        return true;
    }
}

File: 1286-IteratorforCombination/1286-IteratorforCombination.py
"""

1286. Iterator for Combination

Design the CombinationIterator class:

    CombinationIterator(string characters, int combinationLength) Initializes the object with a string characters of sorted distinct lowercase English letters and a number combinationLength as arguments.
    next() Returns the next combination of length combinationLength in lexicographical order.
    hasNext() Returns true if and only if there exists a next combination.

Example 1:

Input
["CombinationIterator", "next", "hasNext", "next", "hasNext", "next", "hasNext"]
[["abc", 2], [], [], [], [], [], []]
Output
[null, "ab", true, "ac", true, "bc", false]

Explanation
CombinationIterator itr = new CombinationIterator("abc", 2);
itr.next();    // return "ab"
itr.hasNext(); // return True
itr.next();    // return "ac"
itr.hasNext(); // return True
itr.next();    // return "bc"
itr.hasNext(); // return False

Constraints:

    1 <= combinationLength <= characters.length <= 15
    All the characters of characters are unique.
    At most 104 calls will be made to next and hasNext.
    It is guaranteed that all calls of the function next are valid.

"""

class CombinationIterator:

    def __init__(self, characters: str, combinationLength: int):
        self.res = []
        def dfs(curr, idx):
            if len(curr) == combinationLength:
                self.res.append(curr)
                return
            for i in range(idx, len(characters)):
                dfs(curr + characters[i], i + 1)
        dfs("", 0)
        self.res = self.res[::-1]

    def next(self) -> str:
        return self.res.pop()

    def hasNext(self) -> bool:
        return len(self.res) > 0
        


# Your CombinationIterator object will be instantiated and called as such:
# obj = CombinationIterator(characters, combinationLength)
# param_1 = obj.next()
# param_2 = obj.hasNext()
